levels/fullstack.tex

```diff
@@ -0,0 +1,251 @@
+    efficient or secure, but correct.
+    &
+
+    Understands basics of web security (can configure TLS on a standalone web
+    server like nginx).
+    &
+
+    Reports progress on a regular basis as required by the team's operational
+    requirements. Actively solicits feedback. Participates on interview panels.
+    &
+
+    Understands and reasons about browser networking concepts. Understands
+    common networking issues and can troubleshoot issues using the developer
+    console.
+    &
+
+    Understands and can write HTML.
+    &
+
+    Understands and can use build tools (like npm and yarn), Docker, and
+    Makefiles.
+    \\ [13em]
+  \hline
+    2&
+
+    \cellcolor{blue!25}
+    Can write high quality user and product focused documentation.
+    &
+
+    &
+
+    Understands how to utilize browser security features (like headers and
+    cookie attributes) to secure web applications.
+    &
+
+    Provides constructive review on peers' code and design. Helps new team
+    members during their first weeks.
+    &
+
+    Has more granular understanding of the networking design, for example can
+    reason about using gRPC vs HTTP/JSON and websockets and their networking
+    and scalability trade-offs. Understands mTLS.
+    &
+
+    Understands how to utilize Javascript, CSS, and DOM knowledge to write
+    frontend applications.
+    &
+
+    Understands and can use transpilers (TypeScript to Javascript) and SPA
+    frameworks (React, Angular). Can leverage components and styles from
+    third-party libraries to implement designs.
+    \\ [12em]
+  \hline
+    3&
+
+    \cellcolor{blue!25}
+    Collaborates with the team to scope requirements, based on good
+    understanding of existing longer term product vision and estimates of the
+    system design of a feature of a product.
+    &
+
+    Understands browser-based event handling and event propagation.
+    Understands the use and implications of closures.
+    &
+
+    Can use frameworks to mitigate common attack vectors for server side (like
+    secure password storage) or client side web applications (like XSS, CSRF, and
+    SSRF) security vulnerabilities.
+    &
+
+    Supports less experienced peers' technical skills, answering questions and
+    being a resource. Documents and improves team practices.
+    &
+
+    Understands scalability and resilience aspects of practical implementation
+    of systems leveraging networking.
+
+    \bigbreak
+
+    Frontend designs are informed by scalable backends, networking techniques
+    such as pagination, connection pooling, backpressure and other aspects of the
+    application design.
+    &
+
+    Can write web applications that have frontend and backend components.
+    Understands web application session and state management techniques.
+
+    \bigbreak
+
+    Senior engineers are not expected to find any faults or provide any further
+    suggestions in the design document submitted by the Level 3 engineer.
```

Comment:
I might be just misunderstanding this point but it seems to be a little too dramatic. Do we expect level 3 devs to submit perfect design documents?

I'd much rather prefer the wording used in the Security column:

> Can build secure systems that will pass quality security audit that will uncover few to no critical system design errors.